tokens/cnft-vault/programs/cnft-vault/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+            account_infos.push(acc.to_account_info());
+        }
+
+        msg!("manual cpi call");
+        solana_program::program::invoke_signed(
+        & solana_program::instruction::Instruction {
+            program_id: ctx.accounts.bubblegum_program.key(),
+            accounts: accounts,
+            data: data,
+        },
+        &account_infos[..],
+        &[&[b"cNFT-vault", &[*ctx.bumps.get("leaf_owner").unwrap()]]])
+        .map_err(Into::into)
+
+        
+    }
+
+    
+    pub fn withdraw_two_cnfts<'info>(ctx: Context<'_, '_, '_, 'info, WithdrawTwo<'info>>,
+        root1: [u8; 32],
+        data_hash1: [u8; 32],
+        creator_hash1: [u8; 32],
+        nonce1: u64,
+        index1: u32,
+        proof_1_length: u8,
+        root2: [u8; 32],
+        data_hash2: [u8; 32],
+        creator_hash2: [u8; 32],
+        nonce2: u64,
+        index2: u32,
+        _proof_2_length: u8 // we don't actually need this (proof_2_length = remaining_accounts_len - proof_1_length)
+    ) -> Result<()> {
+        let merkle_tree1 = ctx.accounts.merkle_tree1.key();
+        let merkle_tree2 = ctx.accounts.merkle_tree2.key();
+        msg!("attempting to send nfts from trees {} and {}", merkle_tree1, merkle_tree2);
+        
+        // TODO check if nft transfers are even valid (correct NFT, correct authority)
+        // in this example anyone can withdraw any NFT from the vault
+
+        let mut accounts1:  Vec<solana_program::instruction::AccountMeta> = vec![
+            AccountMeta::new_readonly(ctx.accounts.tree_authority1.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.leaf_owner.key(), true),
+            AccountMeta::new_readonly(ctx.accounts.leaf_owner.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.new_leaf_owner1.key(), false),
+            AccountMeta::new(ctx.accounts.merkle_tree1.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.log_wrapper.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.compression_program.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.system_program.key(), false),
+        ];
+        
+        let mut accounts2:  Vec<solana_program::instruction::AccountMeta> = vec![
+            AccountMeta::new_readonly(ctx.accounts.tree_authority2.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.leaf_owner.key(), true),
+            AccountMeta::new_readonly(ctx.accounts.leaf_owner.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.new_leaf_owner2.key(), false),
+            AccountMeta::new(ctx.accounts.merkle_tree2.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.log_wrapper.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.compression_program.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.system_program.key(), false),
+        ];
+
+        let transfer_discriminator: [u8;8] = [163, 52, 200, 231, 140, 3, 69, 186];
```

Comment:
Duplicate code, could be a constant on top?

Reply:
yes, please teach me how to do a u8 array as a contant in Rust! XD
(i tried and didn't manage)

Reply:
You couldn't do `const TRANSFER_DISCRIMINATOR: [u8; 8] = ...;` ?